// xbox-rta/dist/common/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertRTAStatus = exports.StatusCode = exports.MessageType = void 0;
var MessageType;
(function (MessageType) {
    MessageType[MessageType["Subscribe"] = 1] = "Subscribe";
    MessageType[MessageType["Unsubscribe"] = 2] = "Unsubscribe";
    MessageType[MessageType["Event"] = 3] = "Event";
    MessageType[MessageType["Resync"] = 4] = "Resync";
})(MessageType || (exports.MessageType = MessageType = {}));
var StatusCode;
(function (StatusCode) {
    StatusCode[StatusCode["Success"] = 0] = "Success";
    StatusCode[StatusCode["UnknownResource"] = 1] = "UnknownResource";
    StatusCode[StatusCode["SubscriptionLimitReached"] = 2] = "SubscriptionLimitReached";
    StatusCode[StatusCode["NoResourceData"] = 3] = "NoResourceData";
    StatusCode[StatusCode["Throttled"] = 1001] = "Throttled";
    StatusCode[StatusCode["ServiceUnavailable"] = 1002] = "ServiceUnavailable";
})(StatusCode || (exports.StatusCode = StatusCode = {}));
const convertRTAStatus = (status) => {
    switch (status) {
        case StatusCode.Success:
            return 'Success';
        case StatusCode.UnknownResource:
            return 'UnknownResource';
        case StatusCode.SubscriptionLimitReached:
            return 'SubscriptionLimitReached';
        case StatusCode.NoResourceData:
            return 'NoResourceData';
        case StatusCode.Throttled:
            return 'Throttled';
        case StatusCode.ServiceUnavailable:
            return 'ServiceUnavailable';
        default:
            return 'Unknown';
    }
};
exports.convertRTAStatus = convertRTAStatus;
